src/basic/cgroup-util.c

```diff
@@ -95,35 +95,39 @@ int cg_enumerate_processes(const char *controller, const char *path, FILE **ret)
         return cg_enumerate_items(controller, path, ret, "cgroup.procs");
 }
 
-int cg_read_pid(FILE *f, pid_t *ret) {
+int cg_read_pid(FILE *f, pid_t *ret, CGroupPidFlags flags) {
         unsigned long ul;
 
         /* Note that the cgroup.procs might contain duplicates! See cgroups.txt for details. */
 
         assert(f);
         assert(ret);
 
-        errno = 0;
-        if (fscanf(f, "%lu", &ul) != 1) {
+        do {
+                errno = 0;
+                if (fscanf(f, "%lu", &ul) != 1) {
 
-                if (feof(f)) {
-                        *ret = 0;
-                        return 0;
+                        if (feof(f)) {
+                                *ret = 0;
+                                return 0;
+                        }
+
+                        return errno_or_else(EIO);
                 }
 
-                return errno_or_else(EIO);
-        }
+                /* In some cirtumstances, cgroups might contain unmappable PIDs from other contexts.
+                 * These show up as zeros, and depending on the caller, can either be plain skipped
+                 * over, or returned as-is for the caller to process them accordingly. */
+        } while (!FLAGS_SET(flags, CGROUP_PID_INCLUDE_UNMAPPED) && ul == 0);
```

Comment:
I misunderstood the code. Please _refuse_ zeros if the flag is set, rather than returning 0 with success. Otherwise, e.g. cg_read_pidref() will provides the pidref of the self process. I thought 0 was skipped or refused. That's why I said REFUSE_ZERO or REFUSE_UNMAPPED is preferable.
Also,
- use `for(;;)` loop rather than `do {} while()`, that's not a strong requirement, but we usually write so.
- cirtumstances -> circumstances

```c
        for(;;) {
                unsigned long ul;

                errno = 0;
                if (fscanf(f, "%lu", &ul) != 1) {

                        if (feof(f)) {
                                *ret = 0;
                                return 0;
                        }

                        return errno_or_else(EIO);
                }

                if (ul > PID_T_MAX)
                        return -EIO;

                /* In some circumstances (e.g. WSL), cgroups might contain unmappable PIDs from other
                 * contexts. These show up as zeros, and depending on the caller, can either be plain skipped
                 * over, or refuse it and return as an error. */
                if (ul <= 0) {
                        if (FLAGS_SET(flags, CGROUP_PID_REFUSE_UNMAPPED))
                                return -EIO;
                        continue;
                }

                *ret = (pid_t) ul;
                return 1;
        }
}
```

Reply:
But refusing the zeros, i.e. keep returning -EIO as is done right now, won't work out for the likes of cg_is_empty(), which would then also return with -EIO instead of stating that the cgroup isn't empty.
Unless it itself is modified to treat -EIO specifically as a special case and just return "not empty" when it encounters it.
Would be happy to change it to that approach if that's preferred. And could then also make cg_migrate() set the flag and just have it fail when it encounters zero-pids, since I still don't think it has a sane way to handle them.

Regarding the loop-style: I found pre-existing examples of all three styles (for(;;), while(){} and do{}while()) in cgroup-util.c alone, and the do{}while-Style was just what required the least amount of changes to the flow of the function.
Also happy to change this to whatever is preferred.